src/features/Details/context/DetailsContext.tsx

```diff
@@ -0,0 +1,86 @@
+import { AppContext } from '@context/AppContext';
+import { Superhero } from '@interfaces/superhero.interface';
+import {
+  createContext,
+  FC,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+
+interface DetailsState {
+  areDetailsLoading: boolean;
+  superhero?: Superhero;
+  fetchDetails: (slug: string) => Promise<Superhero | undefined>;
+}
+
+export const detailsInitialState: DetailsState = {
+  areDetailsLoading: true,
+  superhero: undefined,
+  fetchDetails: () => Promise.resolve(undefined),
+};
+
+export const DetailsContext = createContext<DetailsState>(detailsInitialState);
+
+export const DetailsProvider: FC = ({ children }) => {
+  const delayTimer = useRef<ReturnType<typeof setTimeout>>();
+  const { areHeroesFetched, superheroes } = useContext(AppContext);
+  const [areDetailsLoading, setDetailsLoading] = useState(true);
+  const [superhero, setSuperhero] = useState<Superhero | undefined>();
+
+  const clearTimer = () => {
+    if (delayTimer.current) {
+      clearTimeout(delayTimer.current);
+    }
+  };
+
+  const fetchDetails = (slug: string) =>
+    new Promise<Superhero | undefined>((resolve, reject) => {
+      clearTimer();
+      setDetailsLoading(true);
+
+      // repeat "fetch" if superheroes are not loaded - fake timeout
+      let tries = 0;
+      if (!areHeroesFetched) {
```

Comment:
**suggestion** in this scenario the best approach will be to check if superheroes are already fetched then check if they are fetching and if not try to fetch hero from the `heroes/{id}` endpoint.  

The truth is that the information about all heroes is only necessary for the main view and then it should be reused in other views. In the case of this view, fetching all heroes and filtering by them every time can be a big optimization problem. Especially if the user starts using the application from this place

Reply:
> in this scenario the best approach will be to check if superheroes are already fetched then check if they are fetching and if not try to fetch hero from the `heroes/{id}` endpoint.

@kamil-borecki-cc True. Another thing that I just didn't think of.